modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIoRegistry.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.pagememory.io;
+
+import java.nio.ByteBuffer;
+import java.util.ServiceLoader;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+
+/**
+ * Page IO Registry. This component registers and provides all available {@link PageIo} types.
+ */
+public class PageIoRegistry {
+    /**
+     * Arrays of {@link IoVersions} for fast access. Element 0 is reserved.
+     */
+    private IoVersions<?>[] ioVersions = new IoVersions[65535];
```

Comment:
Should it be `final`?

Reply:
Also, how many registries will we have per Ignite instance? This looks like a suuuuuper-sparse array :)

Reply:
Should be final, you're right. Depending on the future design, there will be either 1 or a small number of registries per node. I could use IntMap or something, but page IO resolving is on hot path